File: example/xx_three/refraction_balls.cpp
module;
#include <infra/Cpp20.h>
#include <xx_three/ex.h>
module two.xxthree;

using namespace two;

EX(xx_refraction_balls)
{
#if UI
	UNUSED(dockbar);
	SceneViewer& viewer = ui::scene_viewer(parent);
	Scene& scene = viewer.m_scene;
	ControlNode& input = viewer;
#else
	static Scene scene = Scene(app.m_gfx);
	static GfxViewer viewer = GfxViewer(window, scene);
	ControlNode& input = window;
#endif

	struct Node { vec3 p; vec3 s; Node3* node; };
	static vector<Node> spheres;

	if(init)
	{
		Camera& camera = viewer.m_camera;
		camera.m_fov = 60.f; camera.m_near = 1.f; camera.m_far = 100000.f;
		camera.m_eye.z = 3200.f;

		Texture& refraction = *app.m_gfx.textures().file("cube/park.jpg.cube");

		Zone& env = scene.m_env;
		env.m_radiance.m_texture = &refraction;
		env.m_radiance.m_filter = false;
		env.m_background.m_texture = &refraction;
		env.m_background.m_mode = BackgroundMode::Panorama;

		env.m_radiance.m_ambient = Colour(1.f);

		Model& sphere = app.m_gfx.shape(Sphere(100.f));

		Program& phong = *app.m_gfx.programs().file("pbr/phong");
		//Program& phong = *app.m_gfx.programs().file("pbr/lambert");

		Material& material = app.m_gfx.materials().create("balls", [&](Material& m) {
			m.m_program = &phong;
			m.m_phong.m_refraction = 0.95f;
		});

		spheres.clear();
		for(size_t i = 0; i < 500; i++)
		{
			vec3 p = vec3(randf(), randf(), randf()) * 10000.f - 5000.f;
			vec3 s = vec3(randf()) * 3.f + 1.f;
			Node3& n = gfx::nodes(scene).add(Node3(p, ZeroQuat, s));
			gfx::items(scene).add(Item(n, sphere, 0U, &material));

			spheres.push_back({ p, s, &n });
		}
	}

	static vec2 mouse = vec2(0.f);
	if(MouseEvent event = input.mouse_event(DeviceType::Mouse, EventType::Moved))
	{
		mouse = (event.m_relative - viewer.m_size / 2.f) * 10.f;
	}

	Camera& camera = viewer.m_camera;
	camera.m_eye.x += (mouse.x - camera.m_eye.x) * .05f;
	camera.m_eye.y += (-mouse.y - camera.m_eye.y) * .05f;

	float timer = app.m_gfx.m_time * -0.01f;
	
	for(size_t i = 0; i < spheres.size(); i++)
	{
		vec3 p = spheres[i].p;
		p.x = cos(timer + float(i)) * 5000.f;
		p.y = sin(timer + float(i) * 1.1f) * 5000.f;
		spheres[i].node->apply(p, ZeroQuat, spheres[i].s);
	}

	//camera.lookAt(scene.position);
}

File: example/xx_three/two/refraction_balls.js
// refraction_balls.js

var viewer = two.ui.scene_viewer(panel);
//two.ui.orbit_controls(viewer);

var camera = viewer.camera;
var scene = viewer.scene;

var zeroq = new two.quat(new two.vec3(0.0));

if(init) {
    this.mouse = new two.vec2(0.0);
    
    camera.fov = 60.0; camera.near = 1.0; camera.far = 100000.0;
    camera.eye.z = 3200.0;

    var refraction = app.gfx.textures.file('cube/park.jpg.cube');
    
    var env = scene.env;
    env.radiance.texture = refraction;
    env.radiance.filter = false;
    env.background.texture = refraction;
    env.background.mode = two.BackgroundMode.Panorama;

    env.radiance.ambient = new two.Colour(1.0);

    var sphere = app.gfx.shape(new two.Sphere(100.0));

    var phong = app.gfx.programs.file('pbr/phong');

    var material = app.gfx.materials.create('balls'); var m = material;
        m.program = phong;
        m.phong.refraction.value = 0.95;
    
    this.spheres = [];
    for(var i = 0; i < 500; i++) {
        
        var p = new two.vec3(Math.random() * 10000.0 - 5000.0, Math.random() * 10000.0 - 5000.0, Math.random() * 10000.0 - 5000.0);
        var s = new two.vec3(Math.random() * 3.0 + 1.0);
        var n = scene.nodes().add(new two.Node3(p, zeroq, s));
        scene.items().add(new two.Item(n, sphere, 0, material));

        this.spheres.push({ p: p, s: s, node: n });
    }
}

var event = viewer.mouse_event(two.DeviceType.Mouse, two.EventType.Moved);
if(event.valid())
{
    this.mouse.x = (event.relative.x - viewer.frame.size.x / 2.0) * 10.0;
    this.mouse.y = (event.relative.y - viewer.frame.size.y / 2.0) * 10.0;
}

var time = app.gfx.time * -0.01;

for(var i = 0; i < this.spheres.length; i++) {
    
    var sphere = this.spheres[i];
    var p = sphere.p;
    p.x = Math.cos(time * -0.01 + i) * 5000.0;
    p.y = Math.sin(time * -0.01 + i * 1.1) * 5000.0;
    sphere.node.apply(p, zeroq, sphere.s);
}

var camera = viewer.camera;
camera.eye.x += (this.mouse.x - camera.eye.x) * .05;
camera.eye.y += (-this.mouse.y - camera.eye.y) * .05;
